Feature: PR previews

JonOfUs opened the pull request:
How these workflows work:

#### `pr-preview-build.yml`
- Is executed on each PR change, runs on the PR itself
- Builds the project, uploads as an artifact

#### `pr-preview-deploy.yml`
- Is executed after each run of `pr-preview-build.yml`
- Runs on the `main` branch, so has the necessary permissions to deploy to Pages
- Fetches the uploaded artifact, deploys it to `OpenPodcastAPI/pr-preview`
- Afterwards, pushes a comment to the PR containing a link to the generated preview

##### Why this needs to be separated into two workflows:
When a PR is opened from a fork of `api-specs`, the workflow also runs on the PR itself, that means in the forked repository. This repo has no permissions to access the secrets of `OPA/api-specs`, which are needed for deployment to GH pages. So it needed to be split, so that the deployment is executed on `main`.

##### Secrets needed
The latter workflow needs permissions to write to `OpenPodcastAPI/pr-preview`. This can be achieved by either a GitHub Application Token or a private/public key comb.
- GitHub Application Token: Allows more specific permission granting, but needs an expiration date => we will need to renew it someday (not much work)
- Public/private key: Allows only full control over the `pr-preview` repo, but does not necessarily expire

(both documented [in the action used](https://github.com/peaceiris/actions-gh-pages#%EF%B8%8F-deploy-to-external-repository-external_repository))

**Application tokens** can be created under: User settings > Developer settings > Personal access tokens > Fine-grained tokens. We'll only need access to the `pr-preview` repo and only `read-and-write` access for the `Contents` permission.

**Private/public keys** can be generated locally, e.g. with the command `ssh-keygen -t ed25519 -C "OPA GH Actions Deploy Key" -f gh-pages`. The public key needs to be added to Repo settings > Deploy keys (`pr-preview`) and the public key needs to be added to Repo settings > Secrets and variables > Actions (`api-specs`).

The current configuration expects a private/public key comb, but I can change this quickly.

Edit: I think I prefer deploy keys (since they can be better narrowed down) over ssh keys. Will change this soon.

Fixes #28

JonOfUs commented:
Changed it to a personal token.
Not sure who has the rights to generate a new personal token, but here is how it's done.

To generate a new personal token:
1. Go to https://github.com/settings/tokens
2. Create a new 'fine-grained token'
3. Insert useful name (e.g. `OPA PR-previews deploy token`), set expiration time to 90 days
4. Select 'Only selected repositories' and then select only 'OpenPodcastAPI/pr-previews'
5. Under 'Repository permissions' go to 'Contents' and set the permission to 'Access: Read and write'
6. Click on 'Generate token' and copy the generated token for later use
8. Go to https://github.com/OpenPodcastAPI/api-specs/settings/secrets/actions (only accessible for authorized users)
9. There, create a 'New repository secret'
10. Name: `PR_DEPLOY_TOKEN`, content: copied token

Afterwards, the action will be able to use this token in order to push the generated build files of a PR to the `pr-preview` repo.

keunes commented:
Ah, thanks for the pointer @JonOfUs. I have now 'enrolled' the org for personal access tokens. You should now be able to request/create one.

keunes commented:
Hi @JonOfUs, I created the token and the repository secret. So it should be ready for use.

JonOfUs commented:
I updated some used Actions to the newest version and replaced some script with an equivalent action for better maintainability. Sadly forgot to name the last commit, it introduces the artifact-downloading action that replaces the difficult-to-read script from before.

@georgkrause you can see the workflows in action in this PR: https://github.com/JonOfUs/OPA_api-specs/pull/8

georgkrause commented:
@JonOfUs Is there an easy way to obtain the preview URL?

georgkrause commented on .github/workflows/pr-preview-build.yml:
```diff
@@ -0,0 +1,50 @@
+          poetry run make html
+      
+      # Store PR number so that it can be extracted in Deploy
+      - name: Create PR artifact
+        run: |
+          mkdir -p ./pr
+          echo ${{ github.event.number }} > ./pr/NR
+          mv docs/_build/html ./pr/
```
Is the really the best way to pass simple values from one job to another in Github Actions? Gitlab CI has environment passing, maybe Github Actions has something similar?

JonOfUs replied:
Agreed, I would like a different solution using e.g. env vars more as well - since the file could potentially be used for some kind of code/information injection.
I couldn't find anything like this, especially since both workflows run on different branches - so maybe this isn't possible in GitHub Actions. The documentation even suggests to use text files in artifacts in this way: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#using-data-from-the-triggering-workflow

georgkrause replied:
Alright then just move on and merge, thank you a lot :)

JonOfUs commented:
Yes, after deployment a new comment with the URL is generated (that also contains a timestamp of deployment and a link to the build files in the PR repo)

georgkrause reviewed (approved):
I don't really like how the information is passed from one job to another, but apart from this, it looks good. Thank you for the work, @JonOfUs !
